_posts/2019/2019-02-10-be-careful-assigning-to-has-one-relations.md

```diff
@@ -0,0 +1,110 @@
+---
+title: Be Careful Assigning to has_one Relations
+description: Strange Think
+layout: article
+category: ruby
+image:
+  base: '2019/be-careful-assigning-to-has-one-relations'
+  alt: 'Broken net'
+  source: "https://unsplash.com/photos/IV--3UEiHlI"
+  credit: "Ben Hershey"
+---
+
+For much of the time, when building relationships between models you typically use `has_many` and `belongs_to`. There are some circumstances where a `has_one` relationship is more appropriate.
+
+However, the behaviour of `has_one` has some wrinkles in its behaviour that make it a little trickier to deal with. See the side-effects of the generated `#association=` method [in the Rails documentation](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_one).
+
+
+## When…
+
+…assigning an invalid new object means you lose the relationship with the currently related model.
+
+```ruby
+class Biography < ApplicationRecord
+  belongs_to :person, optional: true
+
+  validates :text, presence: true
+end
+
+class Person < ApplicationRecord
+  has_one :biography
+end
+
+person = Person.create(name: "Andy Croll")
+=> #<Person id: 1, name: "Andy Croll" ...>
+person.create_biography(text: "Writes emails")
+=> #<Biography id: 1, person_id: 1, text: "Writes emails" ...>
+person.biography = Biography.new
+=> #<Biography id: nil, person_id: 1, text: nil ...>
+person.reload.biography
+=> nil
+```
+
+Oh my.
+
+
+## Use…
+
+…a transaction to wrap the deletion of the existing object. Only delete the existing related object if the new object is valid.
+
+```ruby
+class Biography < ApplicationRecord
+  belongs_to :person, optional: true
+
+  validates :text, presence: true
+end
+
+class Person < ApplicationRecord
+  has_one :biography
+
+  def replace_biography(attributes = nil)
+    Person.transaction do
+      biography&.destroy!
+      create_biography!(attributes)
+    end
+  rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotDestroyed
+    biography # returns invalid object
+  end
+end
+
+person = Person.create(name: "Andy Croll")
+=> #<Person id: 1, name: "Andy Croll" ...>
+person.create_biography(text: "Writes emails")
+=> #<Biography id: 1, person_id: 1, text: "Writes emails" ...>
+person.replace_biography
+=> #<Biography id: nil, person_id: 1, text: nil ...>
+person.reload.biography
+=> #<Biography id: 1, person_id: 1, text: "Writes emails" ...>
+```
+
+Boom.
+
+
+## But why?
+
+The `has_one` relation has always had this behaviour. In the first example the assignment (in this case via `#biography=`) to the `has_one` attribute, removes the previously joined model at the database level even if the newly assigned model isn't valid.
```

Comment:
Is this first sentence necessary?

Reply:
No comma after 'attribute'.

Reply:
The "...removes the previously joined model at the database level..." part...is there a more straightforward way to say this? Also, do we want to say 'association' or 'object' instead of 'model'?

Reply:
I think we definitely want to say "model" because it's an Active Record thing.